exercises/src/main/scala/fpinscala/parallelism/Par.scala

```diff
@@ -3,39 +3,213 @@ package fpinscala.parallelism
-  
+
   private case class UnitFuture[A](get: A) extends Future[A] {
     def isDone = true 
     def get(timeout: Long, units: TimeUnit) = get 
     def isCancelled = false 
     def cancel(evenIfRunning: Boolean): Boolean = false 
   }
-  
+
+  /*
+  Par.map2 needs to return Par[SomeType] because it's used in the book
+  (p. 100) in a call: Par.map2(sum(l), sum(r))(_ + _) within a function
+  (sum) which returns Par[Int].
+
+  In its first parameter list, it needs to take two parameters of
+  Par[T1] and Par[T2] because in the above call, it's called with the
+  arguments sum(l) and sum(r), which we've already seen must be of type
+  Par[SomeType].
+
+  In its second parameter list, it needs to take a function of type (T1,
+  T2) => SomeType because it's called with the argument '_ + _' which
+  takes two arguments and returns SomeType.
+
+  So, replacing
+
+    - T1 with A,
+
+    - T2 with B, and
+
+    - SomeType with C,
+
+  We can infer the actual type to be as follows:
+  */
   def map2[A,B,C](a: Par[A], b: Par[B])(f: (A,B) => C): Par[C] = // `map2` doesn't evaluate the call to `f` in a separate logical thread, in accord with our design choice of having `fork` be the sole function in the API for controlling parallelism. We can always do `fork(map2(a,b)(f))` if we want the evaluation of `f` to occur in a separate thread.
     (es: ExecutorService) => {
       val af = a(es) 
       val bf = b(es)
-      UnitFuture(f(af.get, bf.get)) // This implementation of `map2` does _not_ respect timeouts, and eagerly waits for the returned futures. This means that even if you have passed in "forked" arguments, using this map2 on them will make them wait. It simply passes the `ExecutorService` on to both `Par` values, waits for the results of the Futures `af` and `bf`, applies `f` to them, and wraps them in a `UnitFuture`. In order to respect timeouts, we'd need a new `Future` implementation that records the amount of time spent evaluating `af`, then subtracts that time from the available time allocated for evaluating `bf`.
+
+      new Future[C] {
+        def isDone = af.isDone && bf.isDone
+
+        def get(timeout: Long, units: TimeUnit) = {
+          val startTime = System.currentTimeMillis
+          val a = af.get(timeout, units)
+          val elapsedTime = System.currentTimeMillis - startTime
+          val b =
+            bf.get(
+              /*
+              Give the second future less time to complete because the
+              first future took some time.
+              */
+              TimeUnit.MILLISECONDS.convert(timeout, units) - elapsedTime,
+              TimeUnit.MILLISECONDS
+            )
+
+          f(a, b)
+        }
+
+        def get = get(Long.MaxValue, TimeUnit.MILLISECONDS)
```

Comment:
In practice I know it's effectively the same thing, but this should probably be `def get: C = f(fa.get, fb.get)`, it's just as simple and it avoids the management of the timeout, and there might be different performance characteristics in a real implementation.